src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -5424,6 +5429,22 @@ BOOL IsSafeToCallExecutionManager()
            GCStress<cfg_instr_ngen>::IsEnabled();
 }
 
+#ifdef FEATURE_EMULATE_SINGLESTEP
+BOOL HandleSingleStepAfterException(PCONTEXT contextRecord, PEXCEPTION_RECORD exceptionRecord)
+{
+    Thread *pThread = GetThreadNULLOk();
+    if (pThread && pThread->IsSingleStepEnabled() &&
```

Comment:
Any reason why we only invoke pThread->HandleSingleStep() in some cases and not others? Could we do something like this in IsSafeToHandleHardwareException() and remove the HandleSingleStep call from HandleHardwareException? 

```
#ifdef FEATURE_EMULATE_SINGLESTEP
Thread *pThread = GetThreadNULLOk();
if (pThread != NULL && g_pDebugInterface != NULL)
{
    HandleSingleStep(contextRecord, exceptionRecord, pThread);
}
#endif
```

It feels easier to understand if we always invoke it from one place rather than conditionally invoke it from two different place.